File: compiler/src/feedback/render.ts
import dedent from "dedent";
import type { Node } from "../node";
import type { Type } from "../typecheck";
import { displayType } from "../typecheck";
import type { ResolvedBound } from "../typecheck/constraints/bound";
import { displayBound } from "../typecheck/constraints/bound";
import { code } from "../util/color";

export interface RenderedFeedback {
    strings: readonly string[];
    values: Renderable[];
    render: () => string;
}

export const render = (strings: readonly string[], ...values: Renderable[]): RenderedFeedback => ({
    strings,
    values,
    render: () =>
        dedent(
            values.map((value, index) => strings[index] + value.render()).join("") +
                strings[strings.length - 1],
        ),
});

export abstract class Renderable {
    abstract render(): string;
}

export class RenderableString extends Renderable {
    value: string;

    constructor(value: string) {
        super();
        this.value = value;
    }

    render() {
        return this.value;
    }
}

render.string = (value: string) => new RenderableString(value);

export class RenderableNumber extends Renderable {
    value: number;
    singular: string;
    plural: string;

    constructor(value: number, singular: string, plural: string) {
        super();
        this.value = value;
        this.singular = singular;
        this.plural = plural;
    }

    render() {
        return `${this.value} ${this.value === 1 ? this.singular : this.plural}`;
    }
}

render.number = (value: number, singular: string, plural: string) =>
    new RenderableNumber(value, singular, plural);

export class RenderableNode extends Renderable {
    node: Node;

    constructor(node: Node) {
        super();
        this.node = node;
    }

    render() {
        return this.node.render();
    }
}

render.node = (node: Node) => new RenderableNode(node);

export class RenderableCode extends Renderable {
    code: string;

    constructor(code: string) {
        super();
        this.code = code;
    }

    render() {
        return code(this.code);
    }
}

render.code = (code: string) => new RenderableCode(code);

export class RenderableType extends Renderable {
    type: Type;

    constructor(type: Type) {
        super();
        this.type = type;
    }

    render() {
        return code(displayType(this.type));
    }
}

render.type = (type: Type) => new RenderableType(type);

export class RenderableBound extends Renderable {
    bound: ResolvedBound;

    constructor(bound: ResolvedBound) {
        super();
        this.bound = bound;
    }

    render() {
        return code(displayBound(this.bound));
    }
}

render.bound = (bound: ResolvedBound) => new RenderableBound(bound);

export class RenderableList extends Renderable {
    values: Renderable[];
    separator: string;

    constructor(values: Renderable[], separator: string) {
        super();
        this.values = values;
        this.separator = separator;
    }

    render() {
        const values = this.values.map((value) => value.render());

        if (values.length > 2) {
            return (
                values.slice(0, values.length - 1).join(`, `) +
                `, ${this.separator} ` +
                values[values.length - 1].toString()
            );
        } else if (values.length === 2) {
            return values.join(` ${this.separator} `);
        } else {
            return values[0].toString();
        }
    }
}

render.list = (values: Renderable[], separator: string) => new RenderableList(values, separator);

export type Links = Record<string, Renderable | { or: Renderable[] } | { and: Renderable[] }>;

class RenderableComments extends Renderable {
    comments: string[];
    links: Links;
    suffix?: RenderedFeedback;

    constructor(comments: string[], links: Links, suffix?: RenderedFeedback) {
        super();
        this.comments = comments;
        this.links = links;
        this.suffix = suffix;
    }

    render(): string {
        const { comments, links, suffix } = this;

        const string = comments.join("\n");

        const items = string.split(/\[`([^`]+)`\]/);

        const segments: string[] = [];
        const values: Renderable[] = [];
        for (let i = 0; i < items.length - 1; i += 2) {
            segments.push(items[i]);

            const link = links[items[i + 1]];
            if (link == null) {
                values.push(render.code("_"));
                continue;
            }

            if ("and" in link) {
                values.push(render.list(link.and, "and"));
            } else if ("or" in link) {
                values.push(render.list(link.or, "or"));
            } else if (link instanceof Renderable) {
                values.push(link);
            } else {
                link satisfies never;
            }
        }

        segments.push(items[items.length - 1] + (suffix?.render() ?? ""));

        return render(segments, ...values).render();
    }
}

render.comments = (comments: string[], links: Links, suffix?: RenderedFeedback) =>
    new RenderableComments(comments, links, suffix);

class RenderableOptional extends Renderable {
    value: RenderedFeedback | undefined;

    constructor(value: RenderedFeedback | undefined) {
        super();
        this.value = value;
    }

    render(): string {
        return this.value?.render() ?? "";
    }
}

render.optional = (value: RenderedFeedback | undefined) => new RenderableOptional(value);

File: compiler/feedback/render.go
package feedback

import (
	"bytes"
	"fmt"
	"regexp"
	"strings"

	"wipple/colors"
	"wipple/database"
	"wipple/queries"
	"wipple/typecheck"
)

type Render struct {
	db  *database.Db
	buf bytes.Buffer
}

func NewRender(db *database.Db) *Render {
	return &Render{db: db}
}

func (render *Render) WriteString(s string) {
	fmt.Fprintf(&render.buf, "%s", s)
}

func (render *Render) WriteBreak() {
	fmt.Fprintf(&render.buf, "\n\n")
}

func (render *Render) WriteNumber(n int, singular string, plural string) {
	if n == 1 {
		fmt.Fprintf(&render.buf, "%d %s", n, singular)
	} else {
		fmt.Fprintf(&render.buf, "%d %s", n, plural)
	}
}

func (render *Render) WriteNode(node database.Node) {
	fmt.Fprintf(&render.buf, "%s", database.RenderNode(node))
}

func (render *Render) WriteCode(code string) {
	fmt.Fprintf(&render.buf, "%s", colors.Code(code))
}

func (render *Render) WriteType(ty typecheck.Type) {
	// Get the latest type
	if node, ok := ty.(database.Node); ok {
		if fact, ok := database.GetFact[typecheck.TypedFact](node); ok && fact.Group != nil {
			ty = fact.Group.Types[0]
		}
	}

	fmt.Fprintf(&render.buf, "%s", colors.Code(typecheck.DisplayType(ty, true)))
}

func (render *Render) WriteBound(bound typecheck.ResolvedBound) {
	fmt.Fprintf(&render.buf, "%s", colors.Code(typecheck.DisplayResolvedBound(bound)))
}

func (render *Render) WriteList(items []func(), separator string) {
	if len(items) > 2 {
		for i, item := range items {
			if i > 0 && i == len(items)-1 {
				fmt.Fprintf(&render.buf, ", %s ", separator)
			} else if i > 0 {
				fmt.Fprintf(&render.buf, ", ")
			}

			item()
		}
	} else if len(items) == 2 {
		items[0]()
		fmt.Fprintf(&render.buf, " %s ", separator)
		items[1]()
	} else {
		items[0]()
	}
}

type Links = map[string]func(*Render)

func (render *Render) WriteComments(data queries.CommentsData) {
	links := Links{}
	for name, linksData := range data.Links {
		links[name] = func(render *Render) {
			switch ty := linksData.Type.(type) {
			case database.Node:
				render.WriteNode(ty)
			case *typecheck.ConstructedType:
				render.WriteType(ty)
			}
		}

		links[name+"@related"] = func(render *Render) {
			related := make([]func(), 0, len(linksData.Related))
			for _, link := range linksData.Related {
				related = append(related, func() {
					render.WriteNode(link)
				})
			}

			render.WriteList(related, "and")
		}

		links[name+"@type"] = func(render *Render) {
			types := make([]func(), 0, len(linksData.Types))
			for _, ty := range linksData.Types {
				types = append(types, func() {
					render.WriteType(ty)
				})
			}

			render.WriteList(types, "or")
		}
	}

	commentsString := strings.Join(data.Comments, "\n")
	matches := regexp.MustCompile("(?s)\\[`([^`]+)`\\]").FindAllStringIndex(commentsString, -1)

	lastIndex := 0
	for _, match := range matches {
		start := match[0]
		end := match[1]

		before := commentsString[lastIndex:start]

		key := commentsString[start:end]
		key = key[2 : len(key)-2] // remove brackets

		lastIndex = end

		render.WriteString(before)

		link, ok := links[key]
		if !ok {
			render.WriteCode("_")
			continue
		}

		link(render)
	}

	render.WriteString(commentsString[lastIndex:])
}

func (render *Render) Finish() string {
	return render.buf.String()
}
